services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/utils/GoalStateUtil.java

```diff
@@ -0,0 +1,251 @@
+                                  .build();
+                          Vpc.VpcState vpcState =
+                              Vpc.VpcState.newBuilder().
+                                      setConfiguration(vpcConfiguration).
+                                      setOperationTypeValue(Common.OperationType.CREATE_VALUE)
+                                  .setOperationType(Common.OperationType.CREATE).
+                                      build();
+                          vpcStateSet.add(vpcState);
+                        }
+                      });
+              // leave a dummy sg value since for now there is no impl for sg
+              SecurityGroup.SecurityGroupConfiguration securityGroupConfiguration =
+                  SecurityGroup.SecurityGroupConfiguration.newBuilder().build();
+              final SecurityGroup.SecurityGroupState securityGroupState =
+                  SecurityGroup.SecurityGroupState.newBuilder()
+                      .setConfiguration(securityGroupConfiguration)
+                      .build();
+              final Goalstate.GoalState goalState =
+                  Goalstate.GoalState.newBuilder()
+                      .addAllPortStates(portStateHashSet)
+                      .addAllSubnetStates(subnetStateSet)
+                      .addAllVpcStates(vpcStateSet)
+                      .addSecurityGroupStates(0, securityGroupState)
+                      .build();
+              goalStateHashMap.put(f.getKey(), goalState);
+            });
+
+    return goalStateHashMap;
+  }
+
+  private void fillSubnetAndVpcToPort(
+      Map<String, InternalSubnetEntityNB> subnetMap,
+      Map<String, VpcEntity> vpcMap,
+      InternalPortEntityNB currentPortEntity,
+      Set<InternalPortEntityNB> portStates) {
+    Set<InternalSubnetEntityNB> subnetEntity1HashSet = new HashSet<>();
+    Set<VpcEntity> vpcEntityHashSet = new HashSet<>();
+
+    for (com.futurewei.alcor.web.entity.port.PortEntity.FixedIp fixedIp :
+        currentPortEntity.getFixedIps()) {
+      final InternalSubnetEntityNB subnetEntity1 = subnetMap.get(fixedIp.getSubnetId());
+      subnetEntity1HashSet.add(subnetEntity1);
+      final VpcEntity vpcEntity = vpcMap.get(subnetEntity1.getVpcId());
+      vpcEntityHashSet.add(vpcEntity);
+    }
+    currentPortEntity.setSubnetEntities(subnetEntity1HashSet);
+    currentPortEntity.setVpcEntities(vpcEntityHashSet);
+    portStates.add(currentPortEntity);
+  }
+
+  public List<List<Goalstateprovisioner.GoalStateOperationReply.GoalStateOperationStatus>>
+      talkToACA(Map<String, Goalstate.GoalState> gss, boolean isFast, int port, boolean isOvs) {
+    if (isOvs) {
+      GoalStateService goalStateService = new OVSGoalStateServiceImpl();
+
+      return gss.entrySet()
+          .parallelStream()
+          .map(
+              e -> {
+                goalStateService.setIp(e.getKey());
+                goalStateService.setGoalState(e.getValue());
```

Comment:
Say gss have multiple goal state messages, if we only have one single goalStateService, its setGoalState   method would overwrite its internal goalstate, which could interference with goal state messages, i.e., no parallelism is supported. Is it right?

Reply:
@haboy52581 This is quite significant. Please take a look.

Reply:
As mentioned above, service IS able to handle multiple in parallel, we would use multiple service instance in thread level to make sure performance and secure, if only use service handle multiple GS messages which is later on done on the blockingstub way of grpc, it would anyway block others due to thread safety

Reply:
changed in 5dee85b per discussion